File: example/python/high_level_motion_example.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import time
import threading
import tty
import termios
import os
import logging
from typing import Optional
import math
import magicdog_python as magicdog

logging.basicConfig(
    level=logging.INFO,  # Minimum log level
    format="%(asctime)s [%(levelname)s] %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
)

running = True
robot = None
high_controller = None


# Get single character input (no echo)
def getch():
    fd = sys.stdin.fileno()
    old_settings = termios.tcgetattr(fd)
    try:
        tty.setraw(sys.stdin.fileno())
        ch = sys.stdin.read(1)
        logging.info(f"Received character: {ch}")

        sys.stdout.write("\r")
        sys.stdout.flush()
    finally:
        termios.tcsetattr(fd, termios.TCSADRAIN, old_settings)
    return ch


# Recovery stand
def recovery_stand():
    global high_controller
    try:
        # Set gait to position control standing
        status = high_controller.set_gait(magicdog.GaitMode.GAIT_STAND_R)
        if status.code != magicdog.ErrorCode.OK:
            logging.error(f"Failed to set position control standing: {status.message}")
        else:
            logging.info("Robot set to position control standing")
    except Exception as e:
        logging.error(f"Error executing position control standing: {e}")


# Force control standing
def balance_stand():
    global high_controller
    try:
        # Set gait to force control standing
        status = high_controller.set_gait(magicdog.GaitMode.GAIT_STAND_B)
        if status.code != magicdog.ErrorCode.OK:
            logging.error(f"Failed to set force control standing: {status.message}")
        else:
            logging.info("Robot set to force control standing")
    except Exception as e:
        logging.error(f"Error executing force control standing: {e}")


# Execute trick
def execute_trick(cmd):
    global high_controller
    try:
        action = get_action(cmd)
        # Execute lie down trick
        status = high_controller.execute_trick(action)
        if status.code != magicdog.ErrorCode.OK:
            logging.error(f"Failed to execute trick: {status.message}")
        else:
            logging.info("Trick executed successfully")
    except Exception as e:
        logging.error(f"Error executing trick: {e}")


# Switch gait to down climb stairs mode
def change_gait_to_down_climb_stairs():
    global high_controller
    try:
        current_gait = high_controller.get_gait()
        if current_gait != magicdog.GaitMode.GAIT_DOWN_CLIMB_STAIRS:
            status = high_controller.set_gait(magicdog.GaitMode.GAIT_DOWN_CLIMB_STAIRS)
            if status.code != magicdog.ErrorCode.OK:
                logging.error(f"Failed to set down climb stairs gait: {status.message}")
                return False

            # Wait for gait switch to complete
            while (
                high_controller.get_gait() != magicdog.GaitMode.GAIT_DOWN_CLIMB_STAIRS
            ):
                time.sleep(0.01)
        logging.info("Gait changed to down climb stairs")
        return True
    except Exception as e:
        logging.error(f"Error changing gait: {e}")
        return False


def print_help():
    logging.info("Key function description:")
    logging.info("")
    logging.info("Gait and Trick Functions:")
    logging.info("  1        Function 1: Recovery stand")
    logging.info("  2        Function 2: Force control standing")
    logging.info("  3        Function 3: Execute trick")
    logging.info("")
    logging.info("Joystick Functions:")
    logging.info("  w        Function w: Move forward")
    logging.info("  a        Function a: Move left")
    logging.info("  s        Function s: Move backward")
    logging.info("  d        Function d: Move right")
    logging.info("  t        Function t: Turn left")
    logging.info("  g        Function g: Turn right")
    logging.info("  x        Function x: Stop movement")
    logging.info("")
    logging.info("Head Position Functions:")
    logging.info("  4        Function 4: Get head position")
    logging.info("  5        Function 5: Set head position")
    logging.info("")
    logging.info("  ?        Function ?: Print help")
    logging.info("  ESC      Exit program")


def send_joystick_command(high_controller, left_x, left_y, right_x, right_y):
    if not change_gait_to_down_climb_stairs():
        return
    joy_command = magicdog.JoystickCommand()
    joy_command.left_x_axis = left_x
    joy_command.left_y_axis = left_y
    joy_command.right_x_axis = right_x
    joy_command.right_y_axis = right_y

    status = high_controller.send_joystick_command(joy_command)
    if status.code != magicdog.ErrorCode.OK:
        logging.error(f"Failed to send joystick command: {status.message}")


def get_current_head_position(high_controller):
    status, euler_angles = high_controller.get_current_head_position()
    if status.code != magicdog.ErrorCode.OK:
        logging.error(
            "Failed to get current head position, code: %s, message: %s",
            status.code,
            status.message,
        )
        return

    logging.info("Head Position Roll Angle: %f", euler_angles.roll)
    logging.info("  Pitch Angle: %f", euler_angles.pitch)
    logging.info("  Yaw Angle: %f", euler_angles.yaw)


def set_head_position(high_controller, euler_angles):
    status = high_controller.set_head_position(euler_angles)
    if status.code != magicdog.ErrorCode.OK:
        logging.error(
            "Failed to set head position, code: %s, message: %s",
            status.code,
            status.message,
        )
        return


def deg2rad(deg):
    """角度转弧度"""
    return deg * math.pi / 180.0


def signal_handler(signum, frame):
    """Signal handler function for graceful exit"""
    global robot, running
    running = False
    logging.info("Received interrupt signal (%s), exiting...", signum)
    if robot:
        robot.shutdown()
        logging.info("Robot shutdown")
    exit(-1)


def get_action(cmd):
    if cmd == "102":
        return magicdog.TrickAction.ACTION_LIE_DOWN
    elif cmd == "103":
        return magicdog.TrickAction.ACTION_RECOVERY_STAND
    elif cmd == "33":
        return magicdog.TrickAction.ACTION_SHAKE_HEAD
    else:
        return magicdog.TrickAction.ACTION_NONE


def main():
    """Main function"""
    print_help()

    global robot, high_controller, running

    logging.info("MagicDog SDK Python Example Program")

    robot = magicdog.MagicRobot()
    if not robot.initialize("192.168.55.10"):
        logging.error("Initialization failed")
        return

    if not robot.connect():
        logging.error("Connection failed")
        robot.shutdown()
        return

    # Set motion control level to high level
    status = robot.set_motion_control_level(magicdog.ControllerLevel.HIGH_LEVEL)
    if status.code != magicdog.ErrorCode.OK:
        logging.error(f"Failed to set motion control level: {status.message}")
        robot.shutdown()
        return

    # Get high level motion controller
    high_controller = robot.get_high_level_motion_controller()

    logging.info("Program started, please use keys to control robot...")

    while running:
        logging.info("Enter command: ")
        key = getch()
        if key == "\x1b":  # ESC key
            break

        # 1. Gait and Trick Functions
        # 1.1 Recovery stand
        if key == "1":
            recovery_stand()
        # 1.2 Force control standing
        elif key == "2":
            balance_stand()
        # 1.3 Execute trick
        elif key == "3":
            str_input = input("Enter parameters: ").strip()
            if not str_input:
                continue
            # Split input parameters by space
            parts = str_input.strip().split()

            # Parse parameters
            cmd = parts[0] if parts else "102"
            logging.info(f"Execute trick: {cmd}")
            execute_trick(cmd)
        # 2. Joystick Functions
        # 2.1 Move forward
        elif key.lower() == "w":
            left_y = 1.0
            left_x = 0.0
            right_x = 0.0
            right_y = 0.0
            send_joystick_command(high_controller, left_x, left_y, right_x, right_y)
        # 2.2 Move backward
        elif key.lower() == "s":
            left_y = -1.0
            left_x = 0.0
            right_x = 0.0
            right_y = 0.0
            send_joystick_command(high_controller, left_x, left_y, right_x, right_y)
        # 2.3 Move left
        elif key.lower() == "a":
            left_x = -1.0
            left_y = 0.0
            right_x = 0.0
            right_y = 0.0
            send_joystick_command(high_controller, left_x, left_y, right_x, right_y)
        # 2.4 Move right
        elif key.lower() == "d":
            left_x = 1.0
            left_y = 0.0
            right_x = 0.0
            right_y = 0.0
            send_joystick_command(high_controller, left_x, left_y, right_x, right_y)
        # 2.5 Turn left
        elif key.lower() == "t":
            left_x = 0.0
            left_y = 0.0
            right_x = -1.0
            right_y = 0.0
            send_joystick_command(high_controller, left_x, left_y, right_x, right_y)
        # 2.6 Turn right
        elif key.lower() == "g":
            left_x = 0.0
            left_y = 0.0
            right_x = 1.0
            right_y = 0.0
            send_joystick_command(high_controller, left_x, left_y, right_x, right_y)
        # 2.7 Stop movement
        elif key.lower() == "x":
            left_x = 0.0
            left_y = 0.0
            right_x = 0.0
            right_y = 0.0
            send_joystick_command(high_controller, left_x, left_y, right_x, right_y)
        # 3. Head Position Functions
        # 3.1 Get current head position
        elif key == "4":
            get_current_head_position(high_controller)
        # 3.2 Set head position
        elif key == "5":
            # 处理头部位置设置命令
            try:
                # 获取用户输入
                user_input = input(
                    "请输入三个角度（roll pitch yaw，以空格分隔，-60度到60度）: "
                )
                values = user_input.split()
                if len(values) != 3:
                    logging.error("错误：需要输入三个数字")
                    return
                # 转换为浮点数
                roll_deg = float(values[0])
                pitch_deg = float(values[1])
                yaw_deg = float(values[2])
                # 验证角度范围
                if (
                    not (-60 <= roll_deg <= 60)
                    or not (-60 <= pitch_deg <= 60)
                    or not (-60 <= yaw_deg <= 60)
                ):
                    logging.error("警告：角度应在 -60 到 60 度范围内")
                # 转换为弧度
                euler_angles = magicdog.EulerAngles()
                euler_angles.roll = deg2rad(roll_deg)
                euler_angles.pitch = deg2rad(pitch_deg)
                euler_angles.yaw = deg2rad(yaw_deg)
                # 输出结果
                logging.info("\n转换结果：")
                logging.info(f"Roll:  {roll_deg}° = {euler_angles.roll:.6f} rad")
                logging.info(f"Pitch: {pitch_deg}° = {euler_angles.pitch:.6f} rad")
                logging.info(f"Yaw:   {yaw_deg}° = {euler_angles.yaw:.6f} rad")
                # 设置头部位置
                set_head_position(high_controller, euler_angles)
            except ValueError:
                logging.error("错误：请输入有效的数字")
            except Exception as e:
                logging.error(f"发生错误：{e}")
        # Help
        elif key.upper() == "?":
            print_help()
        else:
            logging.info(f"Unknown key: {key}")

    # Shutdown high level motion controller
    high_controller.shutdown()
    logging.info("High level motion controller shutdown")

    # Disconnect from robot
    robot.disconnect()
    logging.info("Robot disconnected")

    # Shutdown robot
    robot.shutdown()
    logging.info("Robot shutdown")


if __name__ == "__main__":
    main()

File: example/cpp/high_level_motion_example/high_level_motion_example.cpp
#include "magic_robot.h"

#include <termios.h>
#include <unistd.h>
#include <csignal>

#include <iostream>
#include <thread>

using namespace magic::dog;

magic::dog::MagicRobot robot;
std::atomic<bool> is_running(true);

std::atomic<float> left_x_axis(0.0);
std::atomic<float> left_y_axis(0.0);
std::atomic<float> right_x_axis(0.0);
std::atomic<float> right_y_axis(0.0);

void signalHandler(int signum) {
  std::cout << "Interrupt signal (" << signum << ") received.\n";
  is_running.store(false);

  robot.Shutdown();
  // 退出进程
  exit(signum);
}

void print_help(const char* prog_name) {
  std::cout << "按键功能演示程序\n\n";
  std::cout << "用法: " << prog_name << "\n";
  std::cout << "按键功能说明:\n";
  std::cout << "  ESC      退出程序\n";
  std::cout << "  1        功能1:位控站立\n";
  std::cout << "  2        功能2:力控站立\n";
  std::cout << "  3        功能3:执行特技-趴下\n";
  std::cout << "  w        功能4:向前移动\n";
  std::cout << "  a        功能5:向左移动\n";
  std::cout << "  x        功能6:向后移动\n";
  std::cout << "  s        功能7:停止移动\n";
  std::cout << "  d        功能7:向右移动\n";
  std::cout << "  t        功能8:左转向\n";
  std::cout << "  g        功能9:右转向\n";
}

int getch() {
  struct termios oldt, newt;
  int ch;
  tcgetattr(STDIN_FILENO, &oldt);  // 获取当前终端设置
  newt = oldt;
  newt.c_lflag &= ~(ICANON | ECHO);  // 关闭缓冲和回显
  tcsetattr(STDIN_FILENO, TCSANOW, &newt);
  ch = getchar();                           // 读取按键
  tcsetattr(STDIN_FILENO, TCSANOW, &oldt);  // 恢复设置
  return ch;
}

void RecoveryStand() {
  // 获取高层运控控制器
  auto& controller = robot.GetHighLevelMotionController();

  // 设置步态
  auto status = controller.SetGait(GaitMode::GAIT_STAND_R);
  if (status.code != ErrorCode::OK) {
    std::cerr << "set robot gait failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    return;
  }
}

void BalanceStand() {
  // 获取高层运控控制器
  auto& controller = robot.GetHighLevelMotionController();

  // 设置姿态展示步态
  auto status = controller.SetGait(GaitMode::GAIT_STAND_B);
  if (status.code != ErrorCode::OK) {
    std::cerr << "set robot gait failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    return;
  }
  std::cout << "robot gait set to GAIT_BALANCE_STAND successfully." << std::endl;
}

void ExecuteTrick() {
  // 获取高层运控控制器
  auto& controller = robot.GetHighLevelMotionController();

  // 执行特技
  auto status = controller.ExecuteTrick(TrickAction::ACTION_LIE_DOWN);
  if (status.code != ErrorCode::OK) {
    std::cerr << "execute robot trick failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    return;
  }
  std::cout << "robot trick executed successfully." << std::endl;
}

void JoyStickCommand(float left_x_axis,
                     float left_y_axis,
                     float right_x_axis,
                     float right_y_axis) {
  ::left_x_axis.store(left_x_axis);
  ::left_y_axis.store(left_y_axis);
  ::right_x_axis.store(right_x_axis);
  ::right_y_axis.store(right_y_axis);
}

void JoyThread() {
  auto& controller = robot.GetHighLevelMotionController();
    while (is_running.load()) {
      JoystickCommand joy_command;
      joy_command.left_x_axis = left_x_axis;
      joy_command.left_y_axis = left_y_axis;
      joy_command.right_x_axis = right_x_axis;
      joy_command.right_y_axis = right_y_axis;
      auto status = controller.SendJoyStickCommand(joy_command);
      if (status.code != ErrorCode::OK) {
        std::cerr << "send joystick command failed"
                  << ", code: " << status.code
                  << ", message: " << status.message << std::endl;
      }
      usleep(10000);
    }
}

int main(int argc, char* argv[]) {
  // 绑定 SIGINT（Ctrl+C）
  signal(SIGINT, signalHandler);

  std::string local_ip = "192.168.54.111";
  // 配置本机网线直连机器的IP地址，并进行SDK初始化
  if (!robot.Initialize(local_ip)) {
    std::cerr << "robot sdk initialize failed." << std::endl;
    robot.Shutdown();
    return -1;
  }

  // 设置rpc超时时间为5s
  robot.SetTimeout(5000);

  // 连接机器人
  auto status = robot.Connect();
  if (status.code != ErrorCode::OK) {
    std::cerr << "connect robot failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    robot.Shutdown();
    return -1;
  }

  // 切换运控控制器为底层控制器，默认是高层控制器
  status = robot.SetMotionControlLevel(ControllerLevel::HighLevel);
  if (status.code != ErrorCode::OK) {
    std::cerr << "switch robot motion control level failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    robot.Shutdown();
    return -1;
  }

  std::thread joy_thread(JoyThread);

  print_help(argv[0]);

  std::cout << "按任意键继续 (ESC退出)..."
            << std::endl;

  auto change_gait_to_down_climb_stairs = [](auto& robot) -> bool {
    GaitMode current_gait = GaitMode::GAIT_PASSIVE;
    auto status = robot.GetHighLevelMotionController().GetGait(current_gait);
    if (status.code != ErrorCode::OK) {
        std::cerr << "get robot gait failed"
                << ", code: " << status.code
                << ", message: " << status.message << std::endl;
        return false;
    }
    if (current_gait != GaitMode::GAIT_DOWN_CLIMB_STAIRS) {
        status = robot.GetHighLevelMotionController().SetGait(GaitMode::GAIT_DOWN_CLIMB_STAIRS);
        if (status.code != ErrorCode::OK) {
            std::cerr << "set robot gait failed"
                        << ", code: " << status.code
                        << ", message: " << status.message << std::endl;
            return false;
        }
    } else {
        return true;
    }

    status = robot.GetHighLevelMotionController().GetGait(current_gait);
    if (status.code != ErrorCode::OK) {
        std::cerr << "get robot gait failed"
                    << ", code: " << status.code
                    << ", message: " << status.message << std::endl;
        return false;
    }
    while (current_gait != GaitMode::GAIT_DOWN_CLIMB_STAIRS) {
        usleep(10000);
        status = robot.GetHighLevelMotionController().GetGait(current_gait);
        if (status.code != ErrorCode::OK) {
            std::cerr << "get robot gait failed"
                        << ", code: " << status.code
                        << ", message: " << status.message << std::endl;
            return false;
        }
    }
    return true;
  };
  // 等待用户输入
  while (1) {
    int key = getch();
    if (key == 27)
      break;  // ESC键ASCII码为27

    std::cout << "按键ASCII: " << key << ", 字符: " << static_cast<char>(key) << std::endl;
    switch (key) {
      case '1': {
        RecoveryStand();
        break;
      }
      case '2': {
        BalanceStand();
        break;
      }
      case '3': {
        ExecuteTrick();
        break;
      }
      case 'w': {
        if (!change_gait_to_down_climb_stairs(robot)) {
            std::cerr << "change robot gait to down climb stairs failed" << std::endl;
            break;
        }
        JoyStickCommand(0.0, 1.0, 0.0, 0.0);  // 向前
        break;
      }
      case 'a': {
        if (!change_gait_to_down_climb_stairs(robot)) {
            std::cerr << "change robot gait to down climb stairs failed" << std::endl;
            break;
        }
        JoyStickCommand(-1.0, 0.0, 0.0, 0.0);  // 向左
        break;
      }
      case 'x': {
        if (!change_gait_to_down_climb_stairs(robot)) {
            std::cerr << "change robot gait to down climb stairs failed" << std::endl;
            break;
        }
        JoyStickCommand(0.0, -1.0, 0.0, 0.0);  // 向后
        break;
      }
      case 'd': {
        if (!change_gait_to_down_climb_stairs(robot)) {
            std::cerr << "change robot gait to down climb stairs failed" << std::endl;
            break;
        }
        JoyStickCommand(1.0, 0.0, 0.0, 0.0);  // 向右
        break;
      }
      case 't': {
        if (!change_gait_to_down_climb_stairs(robot)) {
            std::cerr << "change robot gait to down climb stairs failed" << std::endl;
            break;
        }
        JoyStickCommand(0.0, 0.0, -1.0, 0.0);  // 左转
        break;
      }
      case 'g': {
        if (!change_gait_to_down_climb_stairs(robot)) {
            std::cerr << "change robot gait to down climb stairs failed" << std::endl;
            break;
        }
        JoyStickCommand(0.0, 0.0, 1.0, 0.0);  // 右转
        break;
      }
      case 's': {
        if (!change_gait_to_down_climb_stairs(robot)) {
            std::cerr << "change robot gait to down climb stairs failed" << std::endl;
            break;
        }
        JoyStickCommand(0.0, 0.0, 0.0, 0.0);  // 停止
        break;
      }
      default:
        std::cout << "未知按键: " << key << std::endl;
        break;
    }

    usleep(10000);
  }

  is_running.store(false);
  joy_thread.join();

  // 断开与机器人的链接
  status = robot.Disconnect();
  if (status.code != ErrorCode::OK) {
    std::cerr << "disconnect robot failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    robot.Shutdown();
    return -1;
  }

  robot.Shutdown();

  return 0;
}